packages/assets-controllers/src/TokenBalancesController.ts

```diff
@@ -107,526 +98,405 @@ export type AllowedEvents =
+              },
+            );
+          }
+
+          // Initialize tokens from allDetectedTokens
+          const detectedChainTokens = this.#detectedTokens[chainId];
+          if (detectedChainTokens?.[account]) {
+            Object.values(detectedChainTokens[account]).forEach(
+              (token: { address: string }) => {
+                const tokenAddress = token.address as ChecksumAddress;
+                ((d.tokenBalances[account] ??= {})[chainId] ??= {})[
+                  tokenAddress
+                ] = '0x0';
+              },
+            );
           }
         }
       }
-    }
-    if (shouldUpdate) {
-      await this.updateBalances({ chainIds }).catch(console.error);
-    }
-  }
 
-  /**
-   * Get an Ethers.js Web3Provider for the requested chain.
-   *
-   * @param chainId - The chain id to get the provider for.
-   * @returns The provider for the given chain id.
-   */
-  #getProvider(chainId: Hex): Web3Provider {
-    return new Web3Provider(this.#getNetworkClient(chainId).provider);
-  }
-
-  /**
-   * Ensures that the block tracker has the latest block data before performing multicall operations.
-   * This is a temporary fix to ensure that the block number is up to date.
-   *
-   * @param chainId - The chain id to update block data for.
-   */
-  async #ensureFreshBlockData(chainId: Hex): Promise<void> {
-    // Force fresh block data before multicall
-    // TODO: This is a temporary fix to ensure that the block number is up to date.
-    // We should remove this once we have a better solution for this on the block tracker controller.
-    const networkClient = this.#getNetworkClient(chainId);
-    await networkClient.blockTracker?.checkForLatestBlock?.();
-  }
+      // Then update with actual fetched balances where available
+      aggregated.forEach(({ success, value, account, token, chainId }) => {
+        if (success && value !== undefined) {
+          ((d.tokenBalances[account] ??= {})[chainId] ??= {})[token] =
+            toHex(value);
+        }
+      });
+    });
 
-  /**
-   * Internal util: run `balanceOf` for an arbitrary set of account/token pairs.
-   *
-   * @param params - The parameters for the balance fetch.
-   * @param params.chainId - The chain id to fetch balances on.
-   * @param params.pairs - The account/token pairs to fetch balances for.
-   * @returns The balances for the given token addresses.
-   */
-  async #batchBalanceOf({
-    chainId,
-    pairs,
-  }: {
-    chainId: Hex;
-    pairs: { accountAddress: Hex; tokenAddress: Hex }[];
-  }): Promise<MulticallResult[]> {
-    if (!pairs.length) {
-      return [];
+    if (!isEqual(prev, next)) {
+      this.update(() => next);
+
+      aggregated.forEach((r) => {
+        if (r.success && r.token === ZERO_ADDRESS) {
+          this.messagingSystem.call(
+            'AccountTrackerController:updateNativeToken',
+            r.account,
+            r.chainId,
+            r.value?.toString() ?? '0',
+          );
+        }
```

Comment:
Hmm since we are doing multichain performance, do you think we can modify this event to take in a list of chains and values.

This way we only emit 1 event to process all updated native tokens.